ml/jml/training_data.h

```diff
@@ -247,6 +248,8 @@ class Training_Data {
     }
 
     const Dataset_Index & generate_index() const;
+
+    friend Multilabel_Training_Data;
```

Comment:
I needed the data_ and index_ of the *inner*. I can get access to the inherited ones but not of the inner ones. I can get around this by either copying the index or exposing its shared pointer publicly.